### ISSUE TITLE
[Bug]: Segfault after downloading first update

### ISSUE THREAD

**Issue by maxsupermanhd** [flatpak/flatpak#6009]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.15.11

### What Linux distribution are you using?

Arch Linux

### Linux distribution version

Linux tower0 6.12.1-arch1-1 #1 SMP PREEMPT_DYNAMIC Fri, 22 Nov 2024 16:04:27 +0000 x86_64 GNU/Linux

### What architecture are you using?

x86_64

### How to reproduce

Run `flatpak update`

### Expected Behavior

Updates to be installed

### Actual Behavior

`Updating 1/19… ████████████████████ 100%  28.3 MB/s  00:00Segmentation fault (core dumped)`

### Additional Information

[core.flatpak.1000.ee8cc01e50494ac1b4c42fa363435ece.3400.1732689733000000.zst.zip](https://github.com/user-attachments/files/17930153/core.flatpak.1000.ee8cc01e50494ac1b4c42fa363435ece.3400.1732689733000000.zst.zip) (real extension is .zst, .zip is for github)
```
ноя 27 09:38:32 tower0 systemd-coredump[2135]: [🡕] Process 2090 (flatpak) of user 1000 dumped core.
                                                  
                                                  Stack trace of thread 2090:
                                                  #0  0x00007c66dc49961c n/a (libc.so.6 + 0x17f61c)
                                                  #1  0x00006077884950bf n/a (flatpak + 0xe60bf)
                                                  #2  0x00006077884d64da n/a (flatpak + 0x1274da)
                                                  #3  0x000060778846289c n/a (flatpak + 0xb389c)
                                                  #4  0x0000607788464f22 n/a (flatpak + 0xb5f22)
                                                  #5  0x0000607788469dfd n/a (flatpak + 0xbadfd)
                                                  #6  0x00006077884a60bd n/a (flatpak + 0xf70bd)
                                                  #7  0x00006077884ab9ac n/a (flatpak + 0xfc9ac)
                                                  #8  0x00006077884392d3 n/a (flatpak + 0x8a2d3)
                                                  #9  0x0000607788432acc n/a (flatpak + 0x83acc)
                                                  #10 0x00006077883ec88a n/a (flatpak + 0x3d88a)
                                                  #11 0x00007c66dc33fe08 n/a (libc.so.6 + 0x25e08)
                                                  #12 0x00007c66dc33fecc __libc_start_main (libc.so.6 + 0x25ecc)
                                                  #13 0x00006077883ee055 n/a (flatpak + 0x3f055)
                                                  
                                                  Stack trace of thread 2093:
                                                  #0  0x00007c66dc3bb451 n/a (libc.so.6 + 0xa1451)
                                                  #1  0x00007c66dc3bdc60 n/a (libc.so.6 + 0xa3c60)
                                                  #2  0x00007c66dc3beef4 malloc (libc.so.6 + 0xa4ef4)
                                                  #3  0x00007c66dcefcf1b g_realloc (libglib-2.0.so.0 + 0x63f1b)
                                                  #4  0x00007c66dcde5006 n/a (libgio-2.0.so.0 + 0x11b006)
                                                  #5  0x00007c66dcde5093 n/a (libgio-2.0.so.0 + 0x11b093)
                                                  #6  0x00007c66dcef6559 n/a (libglib-2.0.so.0 + 0x5d559)
                                                  #7  0x00007c66dcf59157 n/a (libglib-2.0.so.0 + 0xc0157)
                                                  #8  0x00007c66dcef7287 g_main_loop_run (libglib-2.0.so.0 + 0x5e287)
                                                  #9  0x00007c66dcdddbc4 n/a (libgio-2.0.so.0 + 0x113bc4)
                                                  #10 0x00007c66dcf2a026 n/a (libglib-2.0.so.0 + 0x91026)
                                                  #11 0x00007c66dc3ae39d n/a (libc.so.6 + 0x9439d)
                                                  #12 0x00007c66dc43349c n/a (libc.so.6 + 0x11949c)
                                                  
                                                  Stack trace of thread 2098:
                                                  #0  0x00007c66dc425bb0 ppoll (libc.so.6 + 0x10bbb0)
                                                  #1  0x00007c66dcf59227 n/a (libglib-2.0.so.0 + 0xc0227)
                                                  #2  0x00007c66dcef5a55 g_main_context_iteration (libglib-2.0.so.0 + 0x5ca55)
                                                  #3  0x00007c66d93292fe n/a (libdconfsettings.so + 0x62fe)
                                                  #4  0x00007c66dcf2a026 n/a (libglib-2.0.so.0 + 0x91026)
                                                  #5  0x00007c66dc3ae39d n/a (libc.so.6 + 0x9439d)
                                                  #6  0x00007c66dc43349c n/a (libc.so.6 + 0x11949c)
                                                  
                                                  Stack trace of thread 2092:
                                                  #0  0x00007c66dc425bb0 ppoll (libc.so.6 + 0x10bbb0)
                                                  #1  0x00007c66dcf59227 n/a (libglib-2.0.so.0 + 0xc0227)
                                                  #2  0x00007c66dcef5a55 g_main_context_iteration (libglib-2.0.so.0 + 0x5ca55)
                                                  #3  0x00007c66dcef5ab2 n/a (libglib-2.0.so.0 + 0x5cab2)
                                                  #4  0x00007c66dcf2a026 n/a (libglib-2.0.so.0 + 0x91026)
                                                  #5  0x00007c66dc3ae39d n/a (libc.so.6 + 0x9439d)
                                                  #6  0x00007c66dc43349c n/a (libc.so.6 + 0x11949c)
                                                  
                                                  Stack trace of thread 2091:
                                                  #0  0x00007c66dc4311fd syscall (libc.so.6 + 0x1171fd)
                                                  #1  0x00007c66dcf27f20 g_cond_wait (libglib-2.0.so.0 + 0x8ef20)
                                                  #2  0x00007c66dcebe98c n/a (libglib-2.0.so.0 + 0x2598c)
                                                  #3  0x00007c66dcf2e137 n/a (libglib-2.0.so.0 + 0x95137)
                                                  #4  0x00007c66dcf2a026 n/a (libglib-2.0.so.0 + 0x91026)
                                                  #5  0x00007c66dc3ae39d n/a (libc.so.6 + 0x9439d)
                                                  #6  0x00007c66dc43349c n/a (libc.so.6 + 0x11949c)
                                                  
                                                  Stack trace of thread 2094:
                                                  #0  0x00007c66dc425bb0 ppoll (libc.so.6 + 0x10bbb0)
                                                  #1  0x00007c66dcf59227 n/a (libglib-2.0.so.0 + 0xc0227)
                                                  #2  0x00007c66dcef7287 g_main_loop_run (libglib-2.0.so.0 + 0x5e287)
                                                  #3  0x00007c66dc986674 n/a (libpolkit-agent-1.so.0 + 0x4674)
                                                  #4  0x00007c66dcf2a026 n/a (libglib-2.0.so.0 + 0x91026)
                                                  #5  0x00007c66dc3ae39d n/a (libc.so.6 + 0x9439d)
                                                  #6  0x00007c66dc43349c n/a (libc.so.6 + 0x11949c)
                                                  ELF object binary architecture: AMD x86-64
ноя 27 09:38:32 tower0 systemd[1]: Started Pass systemd-coredump journal entries to relevant user for potential DrKonqi handling.
ноя 27 09:38:32 tower0 systemd[1]: Started Process Core Dump (PID 2134/UID 0).
ноя 27 09:38:32 tower0 systemd[1]: Created slice Slice /system/systemd-coredump.
ноя 27 09:38:32 tower0 systemd[1]: Created slice Slice /system/drkonqi-coredump-processor.
ноя 27 09:38:32 tower0 systemd-coredump[2134]: Process 2090 (flatpak) of user 1000 terminated abnormally with signal 11/SEGV, processing...
ноя 27 09:38:32 tower0 kernel: Code: ff ff 48 89 f8 0f bc c9 f3 a4 c3 0f 1f 00 f3 0f 1e fa 89 f8 62 a1 fd 00 ef c0 25 ff 0f 00 00 3d e0 0f 00 00 0f 87 24 01 00 00 <62> f1 7d 20 74 07 c5 fb 93 c0 85 c0 74 16 0f bc>
ноя 27 09:38:32 tower0 kernel: flatpak[2090]: segfault at 0 ip 00007c66dc49961c sp 00007fffa3c3f098 error 4 in libc.so.6[17f61c,7c66dc33e000+171000] likely on CPU 20 (core 10, socket 0)
ноя 27 09:38:31 tower0 flatpak[2090]: user: Pulled runtime/org.freedesktop.Platform.GL.nvidia-565-57-01/x86_64/1.4 from flathub
```

**Comment by mikolfaro** [comment]
Hi, I think I'm experiencing the same issue.

Flatpak version: 1.15.11

What Linux distribution are you using? Arch Linux

Linux distribution version: Linux cthulhu 6.6.63-1-lts #1 SMP PREEMPT_DYNAMIC Fri, 22 Nov 2024 15:39:56 +0000 x86_64 GNU/Linux

What architecture are you using? x86_64

How to reproduce: Run `flatpak update org.freedesktop.Platform.GL.nvidia-565-57-01`

Actual behavior: `Updating… ████████████████████ 100%  13,6 MB/s  00:00Segmentation fault (core dumped)
`

Additional info
```
Thread 5 (Thread 0x7ffff2bfd6c0 (LWP 120015) "flatpak"):
#0  0x00007ffff73c7bb0 in __GI_ppoll (fds=fds@entry=0x7fffe00014a0, nfds=nfds@entry=1, timeout=<optimized out>, timeout@entry=0x0, sigmask=sigmask@entry=0x0) at ../sysdeps/unix/sysv/linux/ppoll.c:42
#1  0x00007ffff7efb227 in ppoll (__fds=0x7fffe00014a0, __nfds=1, __timeout=0x0, __ss=0x0) at /usr/include/bits/poll2.h:101
#2  g_main_context_poll_unlocked (priority=<optimized out>, context=0x7fffe0000b70, timeout_usec=<optimized out>, fds=0x7fffe00014a0, n_fds=1) at ../glib/glib/gmain.c:4591
#3  g_main_context_iterate_unlocked.isra.0 (context=0x7fffe0000b70, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib/glib/gmain.c:4268
#4  0x00007ffff7e99287 in g_main_loop_run (loop=0x7fffe0000d90) at ../glib/glib/gmain.c:4475
#5  0x00007ffff7928674 in server_thread_func (user_data=0x55555572ba20) at ../polkit/src/polkitagent/polkitagentlistener.c:357
#6  0x00007ffff7ecc026 in g_thread_proxy (data=0x7fffdc003840) at ../glib/glib/gthread.c:892
#7  0x00007ffff735039d in start_thread (arg=<optimized out>) at pthread_create.c:447
#8  0x00007ffff73d549c in __GI___clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78

Thread 4 (Thread 0x7ffff33fe6c0 (LWP 120014) "gdbus"):
#0  0x00007ffff73c7bb0 in __GI_ppoll (fds=fds@entry=0x5555559dc190, nfds=nfds@entry=4, timeout=<optimized out>, timeout@entry=0x0, sigmask=sigmask@entry=0x0) at ../sysdeps/unix/sysv/linux/ppoll.c:42
#1  0x00007ffff7efb227 in ppoll (__fds=0x5555559dc190, __nfds=4, __timeout=0x0, __ss=0x0) at /usr/include/bits/poll2.h:101
#2  g_main_context_poll_unlocked (priority=<optimized out>, context=0x55555571e620, timeout_usec=<optimized out>, fds=0x5555559dc190, n_fds=4) at ../glib/glib/gmain.c:4591
#3  g_main_context_iterate_unlocked.isra.0 (context=0x55555571e620, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib/glib/gmain.c:4268
#4  0x00007ffff7e99287 in g_main_loop_run (loop=0x55555571e720) at ../glib/glib/gmain.c:4475
#5  0x00007ffff7d7fbc4 in gdbus_shared_thread_func (user_data=0x555555716400) at ../glib/gio/gdbusprivate.c:284
#6  0x00007ffff7ecc026 in g_thread_proxy (data=0x55555571e740) at ../glib/glib/gthread.c:892
#7  0x00007ffff735039d in start_thread (arg=<optimized out>) at pthread_create.c:447
#8  0x00007ffff73d549c in __GI___clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78

Thread 3 (Thread 0x7fffeb3fe6c0 (LWP 120013) "gmain"):
#0  0x00007ffff73c7bb0 in __GI_ppoll (fds=fds@entry=0x5555557129e0, nfds=nfds@entry=1, timeout=<optimized out>, timeout@entry=0x0, sigmask=sigmask@entry=0x0) at ../sysdeps/unix/sysv/linux/ppoll.c:42
#1  0x00007ffff7efb227 in ppoll (__fds=0x5555557129e0, __nfds=1, __timeout=0x0, __ss=0x0) at /usr/include/bits/poll2.h:101
#2  g_main_context_poll_unlocked (priority=<optimized out>, context=0x555555712760, timeout_usec=<optimized out>, fds=0x5555557129e0, n_fds=1) at ../glib/glib/gmain.c:4591
#3  g_main_context_iterate_unlocked.isra.0 (context=context@entry=0x555555712760, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib/glib/gmain.c:4268
#4  0x00007ffff7e97a55 in g_main_context_iteration (context=0x555555712760, may_block=may_block@entry=1) at ../glib/glib/gmain.c:4338
#5  0x00007ffff7e97ab2 in glib_worker_main (data=<optimized out>) at ../glib/glib/gmain.c:6537
#6  0x00007ffff7ecc026 in g_thread_proxy (data=0x555555712980) at ../glib/glib/gthread.c:892
#7  0x00007ffff735039d in start_thread (arg=<optimized out>) at pthread_create.c:447
#8  0x00007ffff73d549c in __GI___clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78

Thread 2 (Thread 0x7ffff3bff6c0 (LWP 120012) "pool-spawner"):
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x00007ffff7ec9f20 in g_cond_wait_impl (cond=0x55555570e5e8, mutex=0x55555570e5e0) at ../glib/glib/gthread-posix.c:1007
#2  g_cond_wait (cond=0x55555570e5e8, mutex=0x55555570e5e0) at ../glib/glib/gthread.c:1665
#3  0x00007ffff7e6098c in g_async_queue_pop_intern_unlocked (queue=0x55555570e5e0, wait=1, end_time=-1) at ../glib/glib/gasyncqueue.c:375
#4  0x00007ffff7ed0137 in g_thread_pool_spawn_thread (data=<optimized out>) at ../glib/glib/gthreadpool.c:297
#5  0x00007ffff7ecc026 in g_thread_proxy (data=0x55555570c220) at ../glib/glib/gthread.c:892
#6  0x00007ffff735039d in start_thread (arg=<optimized out>) at pthread_create.c:447
#7  0x00007ffff73d549c in __GI___clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78

Thread 1 (Thread 0x7ffff6171680 (LWP 120009) "flatpak"):
#0  __strlen_avx2 () at ../sysdeps/x86_64/multiarch/strlen-avx2.S:76
#1  0x000055555563a0bf in systemd_unit_name_escape (in=0x0) at ../flatpak/common/flatpak-run.c:809
#2  0x000055555567b4da in flatpak_run_in_transient_unit (app_id=<optimized out>, instance_id=0x0, error=0x7fffffffce00) at ../flatpak/common/flatpak-run.c:868
#3  flatpak_run_add_environment_args.isra.0 (bwrap=0x5555559fa250, app_info_path=0x0, flags=<optimized out>, app_id=<optimized out>, context=0x555555aa6e20, app_id_dir=<optimized out>, previous_app_id_dirs=0x0, per_app_dir_lock_fd=-1, instance_id=0x0, exports_out=0x0, error=0x7fffffffd5a0, cancellable=<optimized out>) at ../flatpak/common/flatpak-run.c:514
#4  0x000055555560789c in apply_extra_data (self=self@entry=0x555555739c20, checkoutdir=checkoutdir@entry=0x5555559dfb70, cancellable=cancellable@entry=0x0, error=error@entry=0x7fffffffd5a0) at ../flatpak/common/flatpak-dir.c:8454
#5  0x0000555555609f22 in flatpak_dir_deploy (self=self@entry=0x555555739c20, origin=origin@entry=0x5555559e67e0 "flathub", ref=ref@entry=0x5555557e7350, checksum_or_latest=checksum_or_latest@entry=0x55555593a790 "5e4d40cab2da58e0a8348232fa463aacfc593f469cb390482aca8931012d23fb", subpaths=subpaths@entry=0x5555558ea3a0, previous_ids=previous_ids@entry=0x0, cancellable=0x0, error=0x7fffffffd5a0) at ../flatpak/common/flatpak-dir.c:8893
#6  0x000055555560edfd in flatpak_dir_deploy_update (self=0x555555739c20, ref=0x5555557e7350, checksum_or_latest=0x55555593a790 "5e4d40cab2da58e0a8348232fa463aacfc593f469cb390482aca8931012d23fb", opt_subpaths=0x5555558ea3a0, opt_previous_ids=<optimized out>, cancellable=0x0, error=0x7fffffffd5a0) at ../flatpak/common/flatpak-dir.c:9309
#7  0x000055555564b0bd in flatpak_dir_update (self=0x555555739c20, no_pull=0, no_deploy=0, no_static_deltas=<optimized out>, allow_downgrade=0, app_hint=0, install_hint=0, state=0x5555558eb280, ref=0x5555557e7350, commit=0x55555593a790 "5e4d40cab2da58e0a8348232fa463aacfc593f469cb390482aca8931012d23fb", opt_subpaths=<optimized out>, opt_previous_ids=0x0, sideload_repo=0x0, require_metadata=0x5555559c0a30, token=0x0, progress=0x5555559f8d60, cancellable=0x0, error=0x7fffffffd5a0) at ../flatpak/common/flatpak-dir.c:10788
#8  _run_op_kind (self=self@entry=0x555555739910, op=op@entry=0x5555558eb280, remote_state=remote_state@entry=0x55555573f3c0, out_needs_prune=out_needs_prune@entry=0x7fffffffd770, out_needs_triggers=out_needs_triggers@entry=0x7fffffffd76c, out_needs_cache_drop=out_needs_cache_drop@entry=0x7fffffffd768, cancellable=0x0, error=0x7fffffffd7a0) at ../flatpak/common/flatpak-transaction.c:4841
#9  0x00005555556509ac in flatpak_transaction_real_run (self=0x555555739910, cancellable=0x0, error=0x7fffffffda68) at ../flatpak/common/flatpak-transaction.c:5276
#10 0x00005555555de2d3 in flatpak_cli_transaction_run (transaction=0x555555739910, cancellable=<optimized out>, error=0x7fffffffda68) at ../flatpak/app/flatpak-cli-transaction.c:1683
#11 0x00005555555d7acc in flatpak_transaction_run (transaction=0x555555739910, cancellable=0x0, error=0x7fffffffda68) at ../flatpak/common/flatpak-transaction.c:4737
#12 flatpak_builtin_update (argc=<optimized out>, argv=<optimized out>, cancellable=0x0, error=0x7fffffffda68) at ../flatpak/app/flatpak-builtins-update.c:259
#13 0x000055555559188a in flatpak_run (argc=<optimized out>, argv=<optimized out>, res_error=0x7fffffffda58) at ../flatpak/app/flatpak-main.c:859
#14 main (argc=<optimized out>, argv=<optimized out>) at ../flatpak/app/flatpak-main.c:964
```

**Comment by smcv** [comment]
This seems like a regression from #5962, cc @swick.

Presumably applying "extra data" (Nvidia proprietary driver, openh264, etc.) can call `flatpak_run_add_environment_args()` with a null instance ID? Which means we need to be prepared to substitute some other non-colliding string?

Because instance IDs are *also* small integers, we can no longer use the pid as-is (collisions would become possible); but we could substitute it into `"pid%d"` and use that.

The other possible solution would be to ensure all callers of `flatpak_run_add_environment_args()` generate a unique non-null instance ID.

**Comment by smcv** [comment]
Regression in 1.15.11, fixed in 1.15.12.
